### PR TITLE
Save/Delete the datasets

### PR DESCRIPTION
**Pull request by clementparizot**
Fixes issues #169 & #170
Added a submenu to the datasets with "open" and "delete"
Added a getName() method to the datasetManager to properly set the name
Added a delete() method
Added a confirmation and warning dialog for the delete

### CODE REVIEW

**Comment by johnzeringue** [comment]
If you're unsure about any of the interfaces behaviors, please document the options and start a discussion here.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public DatasetResource(Dataset dataset, Path path, Path basePath) {
+
+        this.dataset = dataset;
+        this.path = path;
+        this.basePath = basePath;
+    }
+
+    public void setPath(Path path) {
+        this.path = path;
+        dataset.setName(this.path.getFileName().toString());
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: This should not be mutable.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public void open() {
+        if (isOpen()) {
+            return;
+        }
+        try {
+             path = Files.move(path, getOpenPath().resolve(path.getFileName()));
+        } catch (IOException ex) {
+            Logger.getLogger(DatasetResource.class.getName()).log(Level.SEVERE, null, ex);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: Use a constant for the `Logger`.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        Optional<DatasetResource> datasetResource = Optional.empty();
+        for (int i = 0 ; i < datasetResources.size(); i++) {
+            if (datasetResources.get(i).getDataset() == dataset) {
+                datasetResource = Optional.of(datasetResources.get(i));
+                break;
+            }
+        }
+        return datasetResource;
+    }
+}
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java]: What is this class doing for you that a `Collection`/`List` or a basic refactoring wouldn't do better?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * @author parizotclement
+ */
+public class DatasetResource {
+
+    private final Dataset dataset;
+    private Path path;
+    private final Path basePath;
+
+    public DatasetResource(Dataset dataset, Path path, Path basePath) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: `DatasetResource` should generate `path` on its own.

File: src/main/java/org/cirdles/topsoil/app/dataset/TSVDatasetManager.java [hunk 5]
```diff
@@ -163,4 +162,30 @@ public boolean isClosed(Dataset dataset) {
...
+        while (Files.exists(getBasePath().resolve("closed").resolve(fileName + ".tsv"))
+                || Files.exists(getBasePath().resolve("open").resolve(fileName + ".tsv"))) {
+            fileName.append("-copy");
+        }
+
+        return fileName.toString();
+    }
+
+    @Override
+    public void delete(Dataset dataset) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/dataset/TSVDatasetManager.java]: Is this class staying or going?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 6]
```diff
@@ -78,7 +82,10 @@
     @FXML
     TabPane dataTableTabPane;
 
-    private DatasetManager datasetManager;
+    //------------TO BE REMOVED----------------
+    //private DatasetManager datasetManager;
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: ?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    private final Path basePath;
+
+    public DatasetResourceFactory(Path basePath) {
+        this.basePath = basePath;
+    }
+
+    public DatasetResource makeDatasetResource(Dataset dataset, Path path) {
+        return new DatasetResource(dataset, basePath, path);
+    }
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java]: You should make a new `DatasetResource` with just a `Dataset`. That way, it can be stored in SQLite, TSV, the cloud or anywhere else. Yes, I do realize that `Dataset`s can be open or closed. That doesn't change anything but implementation.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java (hunk 3 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java]: I'm not a fan of this class right now. If you are, please respond with your thoughts.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                .map(parentPath -> parentPath.equals(getOpenPath()))
+                .orElse(false);
+    }
+
+    public boolean isClosed() {
+        return Optional.ofNullable(path)
+                .map(Path::getParent)
+                .map(parentPath -> parentPath.equals(getClosedPath()))
+                .orElse(false);
+    }
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: This isn't necessary with `DatasetResource`. We can just have `isOpen`.

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: That's because `DatasetResource` is responsible for assuring the third case (neither open nor closed) never happens.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    public void open() {
+        if (isOpen()) {
+            return;
+        }
+        try {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: Put a blank line here (between 103 and 104) and below in `close`, please. It's hard to read without.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    public void save(Path path) {
+
+        this.path = path;
+
+        try {
+            new TSVDatasetWriter().write(dataset, this.path);
+        } catch (IOException ex) {
+            LOGGER.log(Level.SEVERE, null, ex);
+        }
+    }
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: This shouldn't be exposed.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetManager.java [hunk 11]
```diff
@@ -29,8 +29,14 @@
 
     public boolean isClosed(Dataset dataset);
 
+    public boolean isSaved(Dataset dataset);
+
+    public String getName(String name);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetManager.java]: Why a name?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetManager.java [hunk 12]
```diff
@@ -29,8 +29,14 @@
 
     public boolean isClosed(Dataset dataset);
 
+    public boolean isSaved(Dataset dataset);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetManager.java]: Why?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 13]
```diff
@@ -147,31 +156,84 @@ void saveDataTable() {
...
-        MenuItem datasetMenuItem = dataset.getName()
-                .map(MenuItem::new)
-                .orElseGet(MenuItem::new);
+        Menu datasetPopMenu = dataset.getName()
+                .map(Menu::new)
+                .orElseGet(Menu::new);
+
+        MenuItem openMenuItem = new MenuItem("Open");
+        openMenuItem.setOnAction(event -> {
+            loadDataset(dataset);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Get rid of `loadDataset`. This should be object oriented, not imperative.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 14]
```diff
@@ -147,31 +156,84 @@ void saveDataTable() {
...
+        });
 
-        datasetMenuItem.setOnAction(event -> {
-            TopsoilMainWindow.this.loadDataset(dataset);
+        MenuItem deleteMenuItem = new MenuItem("Delete");
+        deleteMenuItem.setOnAction(event -> {
+            deleteDataset(dataset);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Same.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                confirmationDialog().ifPresent((ButtonType choice) -> {
+                    switch (choice.getText()) {
+                        case "Save":
+                            saveDataTable.accept(true);
+                            break;
+                        case "Cancel":
+                            event.consume();
+                            break;
+                    }
+                });
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: This logic is hard to follow.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @author parizotclement
+ */
+public class CloseTabEventHandler implements EventHandler<Event> {
+
+    private final Tab tab;
+    private final DatasetResourceManager datasetResourceManager;
+    private final Consumer<Boolean> saveDataTable;
+
+    public CloseTabEventHandler(Tab tab, DatasetResourceManager datasetResourceManager, Consumer<Boolean> callback) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: Are three parameters necessary for this constructor?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java (hunk 3 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java]: As I've already said, I feel it is more necessary to have this class instead of just a plain list because its purpose is both to create the directories and to retrieve the datasets at the application's initialization. Yes we could maybe do that in the `TopsoilMainWindow`, but as you've said it is already overloaded, and I think it would be out of place to do it in the `DatasetResourceFactory`. 
What do you think would be better ?

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java (hunk 16 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: Well yes they are. I need the access the Tab to handle the closing event of said Tab. 
I also need to check if the tab is empty or not, and for that I need the DatasetResourceManager. I could do it in the TopsoilMainWindow but you've asked me not to put too much into it.
Same with the callback method.

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: I really didn't look earlier when I wrote this because I was on mobile, but I always assume, particularly with my own code, that three argument constructors are usually a sign that I missed something. 

As a matter of fact, none of them are necessary in this case.

The easiest is the first. `(Tab) event.getTarget()` is a drop-in replacement for `tab`, and it allows the same handler instance to be used for multiple tabs.

The second is a bit trickier. This one requires some refactoring, but the key idea is that the only `DatasetResource`s that you'll ever want to close will be associated with a `TSVTable` that's composing the target tab. To me, this is a sign that `TSVTable`s should deal in `DatasetResource`s instead of `Dataset`s (i.e. `setDataset` => `setDatasetResource`).

Finally, the callback. There are a couple of ways to think about this and solve it. The first is to notice that you only ever use this class once, internalize the callback logic, and reduce overall complexity. The second is to leverage OOP, create an abstract class outlining the general pattern, and use the template pattern to replicate the same functionality here with inheritance.

Again, the point of all this attention to detail is to reduce coupling. You were ready to create three dependencies where we don't need more than one, if any at all. It adds up though, and the better version of the code takes no more time to write.

Lots of public methods, lots of parameters, lots of imports—all are signs that you should be rethinking your design. Does this make sense?

On a practical level, there's very little difference between the original implementation and the one I'm suggesting. They both take about the same time and energy to write. In the longterm, however, the former is going to be _significantly_ more cumbersome as the codebase grows and evolves.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        this.tab = tab;
+        this.datasetResourceManager = datasetResourceManager;
+        this.saveDataTable = callback;
+    }
+
+    @Override
+    public void handle(Event event) {
+        if (tab.getContent() instanceof TSVTable) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: I don't think this works any longer, right?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java (hunk 3 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceManager.java]: Creating directories can be the responsibility of the `DatasetResource`, but I think you're right that there could be a purpose for a class like this in loading up old `DatasetResource`s. The trick is to make sure that it's _only_ coordinating the loading of the resources (maybe have it call something a special constructor or method on `DatasetResourceFactory` rather than parsing files directly). It shouldn't have any sort of add method and shouldn't ever take or give `Dataset`s. Think of it more as a `DatasetResourceLoader` than as a manager. `DatasetResource`s _manage themselves_ after they're created. That's the beauty. They handle all the CRUD operations.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java (hunk 15 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: It's hard to follow because you have a `case` inside of an `if`/`else` inside of an `if`. There are like 12 paths to think through here.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java (hunk 16 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: The callback method is in fact the `saveDataTable()` method from `TopsoilMainWindow`. It is already used elsewhere, won't it be recurrent to redefine the exact same method in another class instead of just calling it ?

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: Do you know what the template pattern is? That might be helpful for understanding my second suggestion.

File: src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ *
+ * @author parizotclement
+ */
+public class CloseTabEventHandler implements EventHandler<Event> {
+
+    private final Consumer<Boolean> saveDataTable;
+
+    public CloseTabEventHandler(Consumer<Boolean> callback) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java]: Do you know what the template pattern is yet?

File: src/main/java/org/cirdles/topsoil/app/TSVTable.java [hunk 19]
```diff
@@ -41,8 +41,8 @@
  */
 public class TSVTable extends TableView<Entry> {
 
-    private Path savePath;
-    private Dataset dataset;
+    private Optional<DatasetResource> datasetResource = Optional.empty();
+    private Optional<Dataset> unsavedDataset = Optional.empty(); //Due to pasteFromClipboard
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: Why can't the data from the clipboard be its own `DatasetResource`? Then you have `FileDatasetResource` and `ClipboardDatasetResource`.

File: src/main/java/org/cirdles/topsoil/app/TSVTable.java [hunk 20]
```diff
@@ -58,14 +58,36 @@ public TSVTable() {
         setRowFactory(new EntryRowFactory());
     }
 
+    //Constructor used by TSVTableTest
     public TSVTable(Path savePath) {
         this();
 
-        this.savePath = savePath;
-        load();
+        load(savePath);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: `savePath` and `load` should be taken out with `DatasetResource`s.

File: src/main/java/org/cirdles/topsoil/app/TSVTable.java [hunk 21]
```diff
@@ -118,12 +140,14 @@ public void clear() {
...
 
-    public void load() {
+    //Method used by TSVTableTest
+    public void load(Path savePath) {
         if (savePath != null) {
             loadFromPath(savePath);
         }
     }
 
+    //Method used by TSVTableTest
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: Then refactor by rewriting the tests? I don't know if this comment is a note to self or a fix?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 22]
```diff
@@ -130,67 +135,144 @@ Tab createTab() {
...
 
-    void createDatasetMenuItem(Dataset dataset) {
-        MenuItem datasetMenuItem = dataset.getName()
-                .map(MenuItem::new)
-                .orElseGet(MenuItem::new);
+    private String setDatasetName(String name) {
+
+        StringBuilder fileName = new StringBuilder(name);
+        while (Files.exists(DATASETS_DIRECTORY.resolve("closed").resolve(fileName + ".tsv"))
+                || Files.exists(DATASETS_DIRECTORY.resolve("open").resolve(fileName + ".tsv"))) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: This is kinda ugly.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 23]
```diff
@@ -130,67 +135,144 @@ Tab createTab() {
...
+        Tab tab = createTab();
         Node content = tab.getContent();
 
         if (content instanceof TSVTable) {
             TSVTable table = (TSVTable) content;
 
-            table.setDataset(dataset);
-            datasetManager.open(dataset);
+            table.setDatasetResource(datasetResourceFactory.makeDatasetResource(dataset));
+            table.getDatasetResource().ifPresent(datasetResource -> datasetResource.open());
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Use the `::` syntax please.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 24]
```diff
@@ -310,8 +387,10 @@ void initializeAndShow(Chart chart, Dataset dataset) {
     }
 
     private void initializeAndShow(JavaScriptChart javaScriptChart) {
-        getCurrentTable().map(TSVTable::getDataset).ifPresent(dataset -> {
-            initializeAndShow(javaScriptChart, dataset);
+        getCurrentTable().ifPresent(table -> {
+            table.getDatasetResource().ifPresent(datasetResource -> {
+                initializeAndShow(javaScriptChart, datasetResource.getDataset());
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Please learn to write this with `map` or `flatMap`.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java (hunk 10 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: I mean `void save(Path path)` not `void save()`.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.cirdles.topsoil.app.dataset.writer.TSVDatasetWriter;
+
+/**
+ *
+ * @author parizotclement
+ */
+public class DatasetResource {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: You should really extract `DatasetResource` as an interface. That'll help clear up separation of concerns.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java [hunk 26]
```diff
@@ -15,22 +15,22 @@
...
  */
-public interface DatasetManager {
+public class DatasetResourceFactory {
 
-    public List<Dataset> getDatasets();
+    private final Path basePath;
 
-    public boolean isOpen(Dataset dataset);
+    public DatasetResourceFactory(Path basePath) {
+        this.basePath = basePath;
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java]: `DatasetReader`s and `DatasetWriter`s should be configurable from here as well, although you should be selective about what's able to be set from the constructor.

File: src/main/java/org/cirdles/topsoil/app/TSVTable.java (hunk 19 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: Because you need a path to create a `DatasetResource`, however data from the clipboard are unsaved and thus do not have a path.
We could automatically save them when pasted, but I'm not convinced of the user-friendliness of such a solution. Also, we'd need to have a `DatasetResourceFactory` inside the TSVTable, but as it need the basePath to be created,  we'd have to add a dependency to `TopsoilMainWindow` to retrieve it.

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: I don't think a `DatasetResource` necessarily has a path, hence my comments about some of the methods there. While some _subtypes_ of `DatasetResource` have paths, I think the basic `DatasetResource` can just do CRUD operations and know nothing about the filesystem.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 27]
```diff
@@ -15,22 +15,23 @@
...
 /**
+ * Each DatasetResource handles all CRUD methods for a given Dataset
  *
- * @author John Zeringue
+ * @author parizotclement
  */
-public interface DatasetManager {
+public interface DatasetResource {
+
+    public Dataset getDataset();
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: You should either do `getDataset`, `deleteDataset`, `saveDataset`, ... or `get`, `delete`, `save`, ...

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java [hunk 28]
```diff
@@ -15,22 +15,23 @@
...
 
-    public List<Dataset> getDatasets();
+    public void delete();
 
-    public boolean isOpen(Dataset dataset);
+    public void save();
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResource.java]: `update` might be better.

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+public class DatasetResourceFactory {
+
+    private final Path basePath;
+
+    public DatasetResourceFactory(Path basePath) {
+        this.basePath = basePath;
+    }
+
+    public DatasetResource makeDatasetResource(Dataset dataset) {
+        return (DatasetResource) new SimpleDatasetResource(dataset, basePath);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceFactory.java]: Why are you casting here?

File: src/main/java/org/cirdles/topsoil/dataset/DatasetResourceLoader.java [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+    private final DatasetResourceFactory datasetResourceFactory;
+
+    public DatasetResourceLoader(Path basePath) {
+
+        datasetResourceFactory = new DatasetResourceFactory(basePath);
+
+        try {
+            Files.createDirectories(basePath);
+        } catch (IOException ex) {
+            Logger.getLogger(DatasetResourceLoader.class.getName()).log(Level.SEVERE, null, ex);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/DatasetResourceLoader.java]: Introduce `LOGGER` as I've done elsewhere.

File: src/main/java/org/cirdles/topsoil/dataset/Dataset.java [hunk 31]
```diff
@@ -33,6 +33,8 @@
 
     public Optional<String> getName();
 
+    public void setName(String name);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/Dataset.java]: I don't like making this mutable. I had plans for making the `DatasetReader`s put the filename as the name.

A better solution than adding this would be to modify `getName` to always return a string and add name to the `SimpleDataset` constructor. Of course, then the readers have to take this into account and might sometimes need to generate a name.

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/dataset/Dataset.java]: Or `SimpleDataset` could generate a name for itself without a `name` argument. Example: `dataset-20150617122300`

**Comment by johnzeringue** [comment]
This PR is obsolete due to #184. @clementparizot, thanks for your efforts!
